Meta/serenity.sh

```diff
@@ -0,0 +1,207 @@
+#!/bin/env bash
+set -e
+
+ARG0=$0
+function print_help()
+{
+  cat <<EOF
+Usage: $ARG0 COMMAND [TARGET] [ARGS...]
+  Supported TARGETs: i686 (default), x86_64, lagom
+  Supported COMMANDs:
+    build:      Compiles the target binaries, [ARGS...] are passed through to ninja
+    install:    Installs the target binary
+    image:      Creates a disk image with the installed binaries
+    run:        TARGET lagom:      Runs the unit tests on the build host
+                All other TARGETs: $ARG0 run [TARGET] [kernel_cmd_line]
+                    Runs the built image in QEMU, and optionally passes the
+                    kernel_cmd_line to the Kernel
+    delete:     Removes the build environment for the TARGET
+    recreate:   Deletes and re-creates the build environment for TARGET
+    rebuild:    Deletes and re-creates the build environment, and compiles for TARGET
+    kaddr2line: $ARG0 kaddr2line TARGET ADDRESS
+                    Resolves the ADDRESS in the Kernel/Kernel binary to a file:line
+    addr2line:  $ARG0 addr2line TARGET BINARY_FILE ADDRESS
+                    Resolves the ADDRESS in BINARY_FILE to a file:line. It will
+                    attempt to find the BINARY_FILE in the appropriate build directory
+
+    rebuild-toolchain: Deletes and re-builds the TARGET's toolchain
+
+  Examples:
+    $ARG0 run i686 smp=on
+        Runs the image in QEMU passing "smp=on" to the kernel command line
+    $ARG0 run
+        Runs the image for the default TARGET i686 in QEMU
+    $ARG0 run lagom
+        Runs the unit tests on the build host
+    $ARG0 kaddr2line i686 0x12345678
+        Resolves the address 0x12345678 in the Kernel binary
+    $ARG0 addr2line i686 WindowServer 0x12345678
+        Resolves the address 0x12345678 in the WindowServer binary
+EOF
+}
+
+die() {
+    >&2 echo "die: $*"
+    exit 1
+}
+
+function usage()
+{
+  >&2 print_help
+  exit 1
+}
+
+CMD=$1;
+[ -n "$CMD" ] || usage
+shift
+if [ "$CMD" = "help" ]; then
+  print_help
+  exit 0
+fi
+
+if [ -n "$1" ]; then
+  TARGET="$1"; shift
+else
+  TARGET="i686"
+fi
+CMD_ARGS=( "$@" )
+CMAKE_ARGS=()
+
+function get_top_dir()
+{
+  git rev-parse --show-toplevel
+}
+
+function is_valid_target()
+{
+  if [ $TARGET = "lagom" ]; then
```

Comment:
```suggestion
  if [ "$TARGET" = lagom ]; then
```

I don't understand why shellcheck isn't complaining about this.

You need to quote `$TARGET` here.

Try running this script with `./serenity.sh run 'a b'` and you will see that not quoting this variable causes problems; it will output:
```
./serenity.sh: line 77: [: too many arguments
Unknown target: a b
Usage: ./serenity.sh COMMAND [TARGET] [ARGS...]
  Supported TARGETs: i686 (default), x86_64, lagom
  Supported COMMANDs:
    build:      Compiles the target binaries, [ARGS...] are passed through to ninja
    install:    Installs the target binary
    image:      Creates a disk image with the installed binaries
    run:        TARGET lagom:      Runs the unit tests on the build host
                All other TARGETs: ./serenity.sh run [TARGET] [kernel_cmd_line]
                    Runs the built image in QEMU, and optionally passes the
                    kernel_cmd_line to the Kernel
    delete:     Removes the build environment for the TARGET
    recreate:   Deletes and re-creates the build environment for TARGET
    rebuild:    Deletes and re-creates the build environment, and compiles for TARGET
    kaddr2line: ./serenity.sh kaddr2line TARGET ADDRESS
                    Resolves the ADDRESS in the Kernel/Kernel binary to a file:line
    addr2line:  ./serenity.sh addr2line TARGET BINARY_FILE ADDRESS
                    Resolves the ADDRESS in BINARY_FILE to a file:line. It will
                    attempt to find the BINARY_FILE in the appropriate build directory

    rebuild-toolchain: Deletes and re-builds the TARGET's toolchain

  Examples:
    ./serenity.sh run i686 smp=on
        Runs the image in QEMU passing "smp=on" to the kernel command line
    ./serenity.sh run
        Runs the image for the default TARGET i686 in QEMU
    ./serenity.sh run lagom
        Runs the unit tests on the build host
    ./serenity.sh kaddr2line i686 0x12345678
        Resolves the address 0x12345678 in the Kernel binary
    ./serenity.sh addr2line i686 WindowServer 0x12345678
        Resolves the address 0x12345678 in the WindowServer binary

```

Reply:
If you run it as `./serenity.sh run '*'` it will get even messier.